redbox-review/redbox.go

```diff
@@ -0,0 +1,293 @@
+// Errors occur if there's an invalid input or if there's
+// difficulty setting up either an s3 or redshift connection.
+func NewRedbox(options NewRedboxOptions) (*Redbox, error) {
+	if options.Schema == "" || options.Table == "" || options.S3Bucket == "" {
+		return nil, errIncompleteArgs
+	}
+
+	if options.AWSKey == "" {
+		options.AWSKey = os.Getenv("AWS_ACCESS_KEY_ID")
+	}
+	if options.AWSPassword == "" {
+		options.AWSPassword = os.Getenv("AWS_SECRET_ACCESS_KEY")
+	}
+
+	if options.S3Region == "" {
+		s3Region, err := s3box.GetRegionForBucket(options.S3Bucket)
+		if err != nil {
+			return nil, err
+		}
+		options.S3Region = s3Region
+	}
+
+	s3Box, err := s3box.NewS3Box(s3box.NewS3BoxOptions{
+		S3Bucket:    options.S3Bucket,
+		AWSKey:      options.AWSKey,
+		AWSPassword: options.AWSPassword,
+		BufferSize:  options.BufferSize,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	redshift, err := options.RedshiftConfiguration.RedshiftConnection()
+	if err != nil {
+		return nil, err
+	}
+
+	if options.NManifests <= 0 {
+		options.NManifests = defaultNManifests
+	}
+
+	return newRedboxInjection(options, s3Box, redshift), nil
+}
+
+// Pack writes a single row of bytes. Currently accepts JSON inputs.
+func (rb *Redbox) Pack(row []byte) error {
+	if rb.isShipped() {
+		return errBoxShipped
+	}
+	if rb.isShippingInProgress() {
+		return errShippingInProgress
+	}
+
+	var tempMap map[string]interface{}
+	if err := json.Unmarshal(row, &tempMap); err != nil {
+		return errInvalidJSONInput
+	}
+	return rb.s3Box.Pack(row)
+}
+
+// Ship ships written data to the destination Redshift table.
+// While shipping is in progress, no other operations are permitted.
+// Ship is transactional, meaning that any returned error means
+// the destination table has remained unchanged.
+func (rb *Redbox) Ship() ([]string, error) {
+	if rb.isShipped() {
+		return nil, errBoxShipped
+	}
+	if rb.isShippingInProgress() {
+		return nil, errShippingInProgress
+	}
+
+	// Kick off the s3-to-Redshift job
+	rb.setShippingInProgress(true)
+	defer func() {
+		rb.setShippingInProgress(false)
+	}()
+
+	manifests, err := rb.s3Box.CreateManifests(rb.manifestSlug(), rb.nManifests)
+	if err != nil {
+		return nil, err
+	}
+	if len(manifests) == 0 { // If no data was written, there's nothing to ship.
+		return nil, errNothingToShip
+	}
+
+	if err := rb.copyToRedshift(manifests); err != nil {
+		return nil, err
+	}
+
+	rb.markShipped()
+	return manifests, nil
+}
+
+// manifestSlug defines a convention for the slug of each manifest file.
+func (rb *Redbox) manifestSlug() string {
+	return fmt.Sprintf("%s_%s_%s", rb.schema, rb.table, time.Now().Format(time.RFC3339))
+}
+
+// copyToRedshift transports data pointed to by the manifests into Redshift.
+// If the truncate flag is present the destination table is first cleared.
+func (rb *Redbox) copyToRedshift(manifests []string) error {
+	tx, err := rb.redshift.Begin()
+	if err != nil {
```

Comment:
question: does it make sense to put rollback in a defer? it means you would have to track the error (or use a named return), but might be worth it?